### DISCUSSION TITLE
How can I extract the dominant color of an Image using ImageMagick EXCLUDING white or transparent pixels?

### DISCUSSION THREAD

**Discussion by futuremotiondev** [ImageMagick/ImageMagick#6474] [Help] [answered]
### ImageMagick version

7.1.1-12 Q16-HDRI x64

### Operating system, version and so on

Windows 10 Pro x64 22H2

### Description

I deal a lot with icon design and often need to sort them by color. So I've set out to create a powershell script to do just that, which wraps ImageMagick.

Here's an example folder:

![gfmBk](https://github.com/ImageMagick/ImageMagick/assets/33441569/d19666c7-6cde-450d-94d3-6dd5f70540b1)

The simple explanation: I want to separate all blue icons into a sub-folder. Just as a note: I've tried both SVG and PNG and receive the exact same results. So it's not a file format issue.

I've tried: 
```
magick "Lunacy 32px Toolbar Button Tool Selected.svg" +dither -colors 5 -define histogram:unique-colors=true -format "%c" histogram:info:
```
as well as

```
magick $file -scale 1x1! -alpha off -format "%[pixel:u.p]" info:
```
and finally:

```
magick $File -format %c -depth 8  histogram:info:
```

The last option works the best and actually returns the correct result for images that are not outlined like this one:

![tSSHh](https://github.com/ImageMagick/ImageMagick/assets/33441569/825575a6-ca3c-43ca-b0cc-04f4389d6713)

But images like this are not working:

![aucPu](https://github.com/ImageMagick/ImageMagick/assets/33441569/e74d4715-9e60-4dfb-bceb-6c7e46890ced)

The image has no white (It's transparent). The histogram is listing transparent pixels only. Hence why I want to disable alpha for this process.

Histogram output:

```
2318: (0,0,0,0) #00000000 none
14: (0,128,255,4) #0080FF04 srgba(0,128,255,0.0156863)
2: (0,128,255,2) #0080FF02 srgba(0,128,255,0.00784314)
6: (0,153,204,5) #0099CC05 srgba(0,153,204,0.0196078)
4: (0,170,255,3) #00AAFF03 srgba(0,170,255,0.0117647)
13: (0,255,255,1) #00FFFF01 srgba(0,255,255,0.00392157)
1: (16,159,223,16) #109FDF10 srgba(16,159,223,0.0627451)
3: (17,153,221,15) #1199DD0F srgba(17,153,221,0.0588235)
2: (18,164,219,14) #12A4DB0E srgba(18,164,219,0.054902)
3: (19,161,227,27) #13A1E31B srgba(19,161,227,0.105882)
1: (20,153,224,25) #1499E019 srgba(20,153,224,0.0980392)
2: (20,154,228,38) #149AE426 srgba(20,154,228,0.14902)
1: (20,157,226,26) #149DE21A srgba(20,157,226,0.101961)
1: (21,149,234,12) #1595EA0C srgba(21,149,234,0.0470588)
1: (21,156,227,36) #159CE324 srgba(21,156,227,0.141176)
1: (21,156,227,72) #159CE348 srgba(21,156,227,0.282353)
2: (21,156,229,49) #159CE531 srgba(21,156,229,0.192157)
1: (21,157,225,60) #159DE13C srgba(21,157,225,0.235294)
Etc. (Cut for length)
```

My thinking is the following:

1. I need to turn alpha off (-alpha off)
2. I need to find a way to instruct ImageMagick to ignore white pixels (since after the alpha was turned off, the once transparent pixels will now be white).
3. Finally utilize some method to calculate the dominant color, obeying the previous rule of no alpha/no white pixels.

I would REALLY appreciate any help on this as I'm completely stuck and want to finish this script.

Thank you so much in advance for any help or replies.

**Comment by fmw42** [comment]
First when you do -alpha off on your second icon, the background seems to be black not white.  You just need to sort the histogram and ignore black.  This is a scripting issue, not one for Imagemagick and will be OS dependent.

Perhaps user snibgo might have a script already to find the dominant color in Windows.   I have one but it is for Unix systems.

**Reply by futuremotiondev** [reply]
Thanks for the reply. I'm not sure how to ignore black in the histogram. I'll give it a shot with powershell.

Hopefully snibgo could chime in and help.

**Comment by snibgo** [comment]
The problem seems simple. I don't know why you are aiming for a complex solution with histograms.

The icons are either blue or black, on a transparent background. The two blue examples have slightly varying shades of blue. Scaling them to a single pixel, and then removing alpha, gives the mean colour. Windows BAT or CMD syntax:

```
magick blueIcon.png -scale "1x1^!" -alpha off -precision 15 txt:

# ImageMagick pixel enumeration: 1,1,0,255,srgb
0,0: (23,157,227)  #179DE3  srgb(9.02152788705463%,61.573991950866%,89.0259617952239%)

magick tickBlue.png -scale "1x1^!" -alpha off -precision 15 txt:
# ImageMagick pixel enumeration: 1,1,0,255,srgb
0,0: (23,157,227)  #179DE3  srgb(9.02036557874609%,61.568371147097%,89.0190952353704%)
```
These two examples give similar (but not identical) values. You could use "-precision 2" to get "srgb(9%,62%,89%)".

I don't use Powershell.

**Reply by futuremotiondev** [reply]
@snibgo Thank you for the response! I'll give the code a try. 

The solution I'm trying to create is as follows:
1. The user selects a bunch of **SVG** icons. (I don't know if the fact that the icons are in SVG format complicates things). If it does I'll convert to PNG first or something.
2. The user inputs a HEX color of the icons they want to separate into a new folder. In this case it's `#179DE3` (I'm handling this UI step with PowerShell)
3. All icons with a primary color of #179DE3 get moved to a new folder. I need accuracy here, or a way to specify a small tolerance if necessary. 

It looks like your code handles step three well. Going to test today.

Thanks again,
Jay

**Comment by snibgo** [comment]
SVG files are text files. Depending on their complexity, you might simply `grep` them for the string `#179DE3`. But an SVG file might mention that colour but never use it.

My commands above say the average colour for both images is `#179DE3` (when rounded to 8 bits per channel). You can parse that result from the text output. Alternatively, use:
```
magick tickBlue.png -scale "1x1^!" -alpha off -define ftxt:format=\H\n ftxt:
```
That gives the simple text:
```
#179DE3
```
I expect that you could use the SVG file as input instead of a PNG file.

**Reply by futuremotiondev** [reply]
Thank you so much for all your help. Converting to PNG as an intermediary step is trivial in PowerShell. I just convert them using rsvg-convert and dump into a temp directory. Then I'll perform the imagemagick command. I tried the command you gave me on an SVG and its output was vastly different than when I used a png - and the hex code was like 10 digits, heh.

> I expect that you could use the SVG file as input instead of a PNG file.

Can you explain how? If I can skip the intermediary step of converting to PNG that would be great.

I've attached some example SVG files to this post for you:


![Lunacy 32px Toolbar Pencil Tool Hover](https://github.com/ImageMagick/ImageMagick/assets/33441569/80b19d34-34c8-48aa-a2d3-80422da014a1)![Lunacy 32px Toolbar Scale Tool Selected](https://github.com/ImageMagick/ImageMagick/assets/33441569/c0077596-a27b-4dad-943e-218f1db81382)![Lunacy 32px Toolbar Avatar Tool Hover](https://github.com/ImageMagick/ImageMagick/assets/33441569/38238263-64e3-40df-92c0-10ef3c4e75ff)![Lunacy 32px Toolbar Avatar Tool Selected](https://github.com/ImageMagick/ImageMagick/assets/33441569/50722a3a-9dd1-4167-9e25-68174ab72625)![Lunacy 32px Toolbar Checkbox Tool Selected](https://github.com/ImageMagick/ImageMagick/assets/33441569/652f29d4-43d8-423d-b58e-5b3e264da67e)![Lunacy 32px Toolbar Component Tool Hover](https://github.com/ImageMagick/ImageMagick/assets/33441569/9363b37b-f047-4e20-ba0d-1f47bf5a5da0)

Thank you again for the help.

Jay

**Comment by snibgo** [comment, marked as answer]
The first two SVG files both contain exactly one colour `fill="#179DE3"`. So `grep` (or some Powershell command that searches a text file for a string) would seem to be a simple and fast solution. No need to convert them to PNG files or use ImageMagick.

IM can read SVG files, like this:
```
magick -background None iconPencil.svg -scale "1x1^!" -alpha off -depth 8 -define ftxt:format=\H\n ftxt:
```
The result is:
```
#179DE3
```

**Reply by futuremotiondev** [reply]
Thanks so much, that one liner worked flawlessly.

**Reply by futuremotiondev** [reply]
@snibgo - Really sorry to resurrect this, but for some reason the code you gave above is no longer working correctly. Here's a screen-grab:

https://github.com/ImageMagick/ImageMagick/assets/33441569/4a630729-3d2e-4e54-9a48-65382a75ae2c

Here is an image that shows the mismatch:

![Mismatch](https://github.com/ImageMagick/ImageMagick/assets/33441569/ac2c6b42-ab91-48e6-913b-c5f06110b852)

Here's the SVG icon I used:

![1632523_circle_pinterest_round icon_social media_social network_icon](https://github.com/ImageMagick/ImageMagick/assets/33441569/085f4bfb-7cf2-4993-b7e2-ec3365577055)

The command was working flawlessly before, I have no idea what happened to cause this behavior. Is there a setting I should check? I tried setting the color profile explicitly to `sRGB` thinking it might be a color profile problem. And just for kicks I set the color profile to `RGB` just to see what happened. In every case `magick` is returning incorrect hex values for my SVG files. 

Did I do something wrong? Can you suggest a fix?

I would really appreciate any help at all.

Jay

**Comment by snibgo** [comment]
The examples in your OP had only one colour that wasn't transparent. I showed how you can find that colour, using "-scale 1x1!", which finds the average colour across the non-transparent pixels.

Your Pinterest SVG file contains two colours that are not transparent. Those colours are #BD081C and #FFFFFF. The average colour is, of course, somewhere between those two colours.

**Comment by snibgo** [comment]
In case it isn't obvious from my previous post: you can ignore the white pixels, if you want, by making them transparent.
```
magick -background None pinterest.svg -fuzz 40% -transparent white -write x.png -scale "1x1^!" -alpha off -depth 8 -define ftxt:format=\H\n ftxt:

#BD081C
```
The large fuzz factor is needed to remove most of the antialiasing.

**Comment by snibgo** [comment]
If you have many solid colours, you can do this:
```
magick -background None pinterest.svg -format %c -depth 8  histogram:info: |grep ,255) |sort
```
"`grep ,255)`" removes any colours that are not full opaque.

The final few lines are:
```
            25: (190,11,31,255) #BE0B1FFF srgba(190,11,31,1)
            26: (254,253,253,255) #FEFDFDFF srgba(254,253,253,1)
            28: (191,14,33,255) #BF0E21FF srgba(191,14,33,1)
            37: (189,9,29,255) #BD091DFF srgba(189,9,29,1)
         26804: (255,255,255,255) #FFFFFFFF white
        176205: (189,8,28,255) #BD081CFF srgba(189,8,28,1)
```
The most common colour is `#BD081CFF`. (The final `FF` can be stripped.)

**Comment by futuremotiondev** [comment]
Thank you so much for the examples. I didn't realize until later that the SVG I supplied had a white fill.

Is there any way possible to combine both of your last two approaches? I.E.

1. Make white pixels transparent
2. Also remove any colors that are not fully opaque
3. Sort the histogram and extract the most dominant color as you did in your last post

Would this be feasible? I never want white pixels to "count" as a color in this use-case. 

Thanks so much again for the help. Let me know if the approach I outlined above is possible.

Jay

**Comment by snibgo** [comment]
Yes, the commands can be combined. For example:
```
magick -background None c:\web\im\pinterest.svg -transparent white -format %c -depth 8  histogram:info: |grep ,255) |sort
```
The last few lines are:
```
            26: (254,253,253,255) #FEFDFDFF srgba(254,253,253,1)
            28: (191,14,33,255) #BF0E21FF srgba(191,14,33,1)
            37: (189,9,29,255) #BD091DFF srgba(189,9,29,1)
        176205: (189,8,28,255) #BD081CFF srgba(189,8,28,1)
```
This does the three tasks you asked for.

The first column  of the output is a count of pixels with the given colour. A script could read this and determine which text rows have trivial numbers of pixels, and hence how many colours have a significant number of pixels. In this case, there is only one colour.

**Reply by futuremotiondev** [reply]
I can't get this command to work verbatim:

`magick -background None c:\web\im\pinterest.svg -transparent white -format %c -depth 8  histogram:info: |grep ,255) |sort`

If I remove `|grep ,255) |sort`, I get a histogram and the command functions. I have a feeling you're on *NIX? Is there an alternative way to structure this command so it functions on Windows?

This is my code in Powershell right now:

```
$Margs1 = "-background", "None"
$File = "C:\Icons\00 Brands\Brands Dev Sets\32 Crypto CDowner\Crypto Flat Bela.svg"
$Margs2 = "-format", "%c", "-depth", "8", 'histogram:info:'
& magick $Margs1 $File $Margs2 
```
Output (I trimmed out a bunch of values for the sake of readability):

```
             168: (0,0,0,0) #00000000 none
             1: (17,159,244,45) #119FF42D srgba(17,159,244,0.176471)
             2: (17,161,244,46) #11A1F42E srgba(17,161,244,0.180392)
             1: (18,159,246,85) #129FF655 srgba(18,159,246,0.333333)
             2: (18,159,247,152) #129FF798 srgba(18,159,247,0.596078)
             2: (18,160,246,140) #12A0F68C srgba(18,160,246,0.54902)
             1: (18,160,246,86) #12A0F656 srgba(18,160,246,0.337255)
             1: (18,160,246,193) #12A0F6C1 srgba(18,160,246,0.756863)
             2: (18,160,247,153) #12A0F799 srgba(18,160,247,0.6)
             2: (18,161,246,87) #12A1F657 srgba(18,161,246,0.341176)
             2: (18,161,246,138) #12A1F68A srgba(18,161,246,0.541176)
             1: (18,161,246,208) #12A1F6D0 srgba(18,161,246,0.815686)
             3: (18,161,247,154) #12A1F79A srgba(18,161,247,0.603922)
             2: (18,164,255,14) #12A4FF0E srgba(18,164,255,0.054902)
             2: (19,159,246,200) #139FF6C8 srgba(19,159,246,0.784314)
             1: (19,160,245,134) #13A0F586 srgba(19,160,245,0.52549)
           480: (19,160,246,255) #13A0F6FF srgba(19,160,246,1)
             6: (19,160,246,249) #13A0F6F9 srgba(19,160,246,0.976471)
             6: (19,160,246,244) #13A0F6F4 srgba(19,160,246,0.956863)
             4: (19,160,246,199) #13A0F6C7 srgba(19,160,246,0.780392)
             4: (19,160,246,250) #13A0F6FA srgba(19,160,246,0.980392)
             4: (19,160,246,234) #13A0F6EA srgba(19,160,246,0.917647)
             4: (19,160,246,233) #13A0F6E9 srgba(19,160,246,0.913725)
             4: (19,160,246,202) #13A0F6CA srgba(19,160,246,0.792157)
             3: (19,160,246,227) #13A0F6E3 srgba(19,160,246,0.890196)
             3: (19,160,246,245) #13A0F6F5 srgba(19,160,246,0.960784)
             2: (19,160,246,248) #13A0F6F8 srgba(19,160,246,0.972549)
             2: (19,160,246,198) #13A0F6C6 srgba(19,160,246,0.776471)
             2: (19,160,246,226) #13A0F6E2 srgba(19,160,246,0.886275)
             2: (19,160,246,251) #13A0F6FB srgba(19,160,246,0.984314)
             1: (19,160,246,243) #13A0F6F3 srgba(19,160,246,0.952941)
             1: (19,160,246,229) #13A0F6E5 srgba(19,160,246,0.898039)
             1: (19,160,246,204) #13A0F6CC srgba(19,160,246,0.8)
             2: (19,160,247,150) #13A0F796 srgba(19,160,247,0.588235)
           146: (255,255,255,255) #FFFFFFFF white
```
So I'm getting the correct output: `480: (19,160,246,255) #13A0F6FF srgba(19,160,246,1)` this color corresponds to the image I used. I just need to find a way to remove everything that's not fully opaque. An eqivilent of `"grep ,255)"` for windows.

**EDIT:**

I got it to work with `findstr` in PowerShell. So output now looks like this (Also trimmed for readability):

```
             480: (19,160,246,255) #13A0F6FF srgba(19,160,246,1)
             1: (20,160,246,255) #14A0F6FF srgba(20,160,246,1)
             2: (21,161,246,255) #15A1F6FF srgba(21,161,246,1)
             1: (23,161,246,255) #17A1F6FF srgba(23,161,246,1)
             2: (24,162,246,255) #18A2F6FF srgba(24,162,246,1)
             1: (25,163,246,255) #19A3F6FF srgba(25,163,246,1)
             2: (26,163,246,255) #1AA3F6FF srgba(26,163,246,1)
             1: (35,166,247,255) #23A6F7FF srgba(35,166,247,1)
             1: (163,218,252,255) #A3DAFCFF srgba(163,218,252,1)
             1: (164,218,252,255) #A4DAFCFF srgba(164,218,252,1)
             1: (171,221,252,255) #ABDDFCFF srgba(171,221,252,1)
             1: (174,223,252,255) #AEDFFCFF srgba(174,223,252,1)
             1: (177,224,252,255) #B1E0FCFF srgba(177,224,252,1)
             1: (180,225,252,255) #B4E1FCFF srgba(180,225,252,1)
             2: (184,226,252,255) #B8E2FCFF srgba(184,226,252,1)
             2: (186,227,252,255) #BAE3FCFF srgba(186,227,252,1)
             1: (187,227,252,255) #BBE3FCFF srgba(187,227,252,1)
             1: (190,229,253,255) #BEE5FDFF srgba(190,229,253,1)
             1: (194,230,253,255) #C2E6FDFF srgba(194,230,253,1)
             1: (195,231,253,255) #C3E7FDFF srgba(195,231,253,1)
             1: (198,232,253,255) #C6E8FDFF srgba(198,232,253,1)
             1: (199,233,253,255) #C7E9FDFF srgba(199,233,253,1)
             2: (203,234,253,255) #CBEAFDFF srgba(203,234,253,1)
             1: (213,238,253,255) #D5EEFDFF srgba(213,238,253,1)
             2: (244,251,255,255) #F4FBFFFF srgba(244,251,255,1)
             1: (245,251,255,255) #F5FBFFFF srgba(245,251,255,1)
             1: (247,252,255,255) #F7FCFFFF srgba(247,252,255,1)
             1: (249,252,255,255) #F9FCFFFF srgba(249,252,255,1)
             1: (249,253,255,255) #F9FDFFFF srgba(249,253,255,1)
             1: (250,253,255,255) #FAFDFFFF srgba(250,253,255,1)
             1: (251,254,255,255) #FBFEFFFF srgba(251,254,255,1)
             1: (252,254,255,255) #FCFEFFFF srgba(252,254,255,1)
             1: (253,254,255,255) #FDFEFFFF srgba(253,254,255,1)
             2: (254,255,255,255) #FEFFFFFF srgba(254,255,255,1)
            146: (255,255,255,255) #FFFFFFFF white
       ```
       
However, one thing that doesn't make sense to me is why `146: (255,255,255,255) #FFFFFFFF white` is in the output. I thought the script made white pixels transparent? If the output isn't wrong, that aspect of the command isn't working right.

**Comment by snibgo** [comment]
> I have a feeling you're on *NIX? 

I am on Windows, mostly using either the command-line or BAT script files. I don't use Powershell, and can't advise on that.

`grep` is available for most platforms, including Windows. A similar Microsoft Windows command is `findstr`. I expect Powershell has the equivalent facility. Similarly `sort`.

> I thought the script made white pixels transparent?

The command I showed includes "`-transparent white`". You don't have that in your first command , and don't show the command you used with `findstr`.

**Comment by futuremotiondev** [comment]
> The command I showed includes "`-transparent white`". You don't have that in your first command , and don't show the command you used with `findstr`.

Sorry for omitting that. I got ALMOST everything working and sorted. This is effectively the command I'm using:

`magick -background None "C:\Wallpapers\wallpaperflare.com_wallpaper3.jpg" -transparent White -format %c -depth 8 histogram:info: | findstr ',255)'`

This outputs all the histogram data properly, and I've successfully parsed it into an organized data format that I can consume in PowerShell.

The biggest problem I'm having now is speed. At first, the histogram generation on one of my images took literally 45–50 seconds and printed about 10,000 lines or more of data when I wrote the output to screen. It **completely** filled up my terminal's buffer (9999 lines) - and then some. 

But when I removed the print command, things sped up a bit. I think the actual print operation to screen was causing a much higher execution time. However, I'm still seeing anywhere between 10-30 second processing times, and it's obvious that the histogram generation is the bottleneck here as I physically observed it generating over 10,000 values.

One file I tried was a 720 KB `.jpg`. Here is the total execution time:

![Code_tHUFj6u2rO](https://github.com/ImageMagick/ImageMagick/assets/33441569/bc4d45a1-f017-427a-b582-3302c0b75409)

This kind of bottleneck is a huge bummer. 

Is there any way at all to speed up the histogram generation? Disregard low intensity values somehow? Resize images that are above a certain WxH threshold? Anything? You can try yourself with the same image if you want. I've attached it to this post.

Don't get me wrong, I really, _really_ appreciate your help. Thank you so much for all you've showed me. 

![wallpaperflare com_wallpaper3](https://github.com/ImageMagick/ImageMagick/assets/33441569/a73eae4c-a86d-4e15-979a-e9d63b0668ce)

**Edit:** Another thing I thought of. Is there a way to effectively tell ImageMagick to skip every `n` pixel during the histogram process? Sampling every 3rd or 4th pixel could render much appreciated performance gains.

**Reply by fmw42** [reply]
Add -depth 8 and then perhaps reduce the number of colors.

**Reply by futuremotiondev** [reply]
Yeah, that's a good call. I think I'm also going to resize all images for this function to a max of 500-500px - 700x700px. That should greatly reduce the amount of pixels being scanned for the histogram and still get me where I need to be.

**Comment by snibgo** [comment]
> But when I removed the print command, things sped up a bit.

What print command? There is no need to write any of the data to the screen. Just pipe it into a command that filters out the lines you don't want.

>  Is there a way to effectively tell ImageMagick to skip every n pixel during the histogram process? Sampling every 3rd or 4th pixel could render much appreciated performance gains.

You can "`-sample 50%`" immediately after reading the input. This will reduce the number of pixels horizontallly and vertically to 50%, so 25% overall. See https://imagemagick.org/script/command-line-options.php#sample

**Comment by futuremotiondev** [comment]
> What print command? There is no need to write any of the data to the screen. Just pipe it into a command that filters out the lines you don't want.

I know this. I had to print the histogram in early development, so I could correctly parse the generated data. There is no more print command anymore.

> You can "-sample 50%" immediately after reading the input. This will reduce the number of pixels horizontallly and vertically to 50%, so 25% overall. See https://imagemagick.org/script/command-line-options.php#sample

This is a fascinating command. I didn't know it existed. My original thought was to resize all images to a maximum of 500-500px - 700x700px, but it looks like this command does essentially the same thing but faster? I'm going to give it a shot. Thanks again for all your help.

**Comment by snibgo** [comment]
> My original thought was to resize all images to a maximum of 500-500px - 700x700px, but it looks like this command does essentially the same thing but faster?

`-resize` and `-scale` and `-sample` are similar. They can all increase or decrease the number of pixels in an image, but they do so in different ways. Suppose a number of input pixels contribute to each output pixel:

1. `-resize` generally uses "elliptical weighting average" of the input pixels, which is complex but gives good results eg on photos. This can introduce new colours.

2. `-scale` will use a simple average of the input pixels. For example, scaling to a single row makes an output where each pixel is the average of the pixels in a column of the input. This can also introduce new colours.

3. `-sample` will pick one of the input pixels, and just use that for the output. This is the fastest method, and reduces quality of photos, but sometimes we want to "skip over input pixels", which this does. This introduces no new colours.

If you are not familiar with IM operations, it is worth skim-reading the https://imagemagick.org/script/command-line-options.php page, to get an idea of what can be done easily.
